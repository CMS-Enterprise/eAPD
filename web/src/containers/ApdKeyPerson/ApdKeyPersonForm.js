import { TextField } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import { t } from '../../i18n';
import Choice from '../../components/Choice';
import PersonCostForm from '../../components/PersonCostForm';

import {
  setKeyPersonCost,
  setKeyPersonEmail,
  setKeyPersonHasCosts,
  setKeyPersonName,
  setKeyPersonFTE,
  setKeyPersonRole
} from '../../actions/editApd';

const tRoot = 'apd.stateProfile.keyPersonnel';

const PersonForm = ({
  index,
  item: { costs, email, hasCosts, name, fte, position },
  setCost,
  setEmail,
  setHasCosts,
  setName,
  setRole,
  setTime,
  years
}) => {
  const handleChange = action => ({ target: { value } }) => {
    action(index, value);
  };

  const setPersonHasCosts = newHasCosts => () => {
    setHasCosts(index, newHasCosts);
  };

  const setCostForYear = (year, value) => {
    setCost(index, year, value);
  };

  const setFTEForYear = (year, value) => {
    setTime(index, year, +value);
  };

  const primary = index === 0;

  return (
    <Fragment>
      <h4 className="ds-h4">
        {primary ? 'Primary' : 'Additional'} APD Point of Contact
      </h4>
      <TextField
        autoFocus
        name={`apd-state-profile-pocname${index}`}
        label={t(`${tRoot}.labels.name`)}
        value={name}
        onChange={handleChange(setName)}
      />
      <TextField
        name={`apd-state-profile-pocemail${index}`}
        label={t(`${tRoot}.labels.email`)}
        value={email}
        onChange={handleChange(setEmail)}
      />
      <TextField
        name={`apd-state-profile-pocposition${index}`}
        label={t(`${tRoot}.labels.position`)}
        value={position}
        onChange={handleChange(setRole)}
      />

      <fieldset className="ds-c-fieldset">
        <legend className="ds-c-label">{t(`${tRoot}.labels.hasCosts`)}</legend>
        <Choice
          type="radio"
          name={`apd-state-profile-hascosts-no${index}`}
          value="no"
          checked={!hasCosts}
          onChange={setPersonHasCosts(false)}
        >
          No
        </Choice>
        <Choice
          type="radio"
          name={`apd-state-profile-hascosts-yes${index}`}
          value="yes"
          checked={hasCosts}
          onChange={setPersonHasCosts(true)}
          checkedChildren={
            <PersonCostForm
              items={years.reduce(
                (o, year) => ({
                  ...o,
                  [year]: {
                    amt: costs[year],
                    perc: fte[year]
                  }
                }),
                {}
              )}
              setCost={setCostForYear}
              setFTE={setFTEForYear}
            />
          }
        >
          Yes
        </Choice>
      </fieldset>
    </Fragment>
  );
};
PersonForm.propTypes = {
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    costs: PropTypes.object.isRequired,
    email: PropTypes.string.isRequired,
    hasCosts: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    fte: PropTypes.object.isRequired,
    position: PropTypes.string.isRequired
  }).isRequired,
  setCost: PropTypes.func.isRequired,
  setEmail: PropTypes.func.isRequired,
  setHasCosts: PropTypes.func.isRequired,
  setName: PropTypes.func.isRequired,
  setRole: PropTypes.func.isRequired,
  setTime: PropTypes.func.isRequired,
  years: PropTypes.array.isRequired
};

const mapDispatchToProps = {
  setCost: setKeyPersonCost,
  setEmail: setKeyPersonEmail,
  setHasCosts: setKeyPersonHasCosts,
  setName: setKeyPersonName,
  setRole: setKeyPersonRole,
  setTime: setKeyPersonFTE
};

export default connect(null, mapDispatchToProps)(PersonForm);

export { PersonForm as plain, mapDispatchToProps };
