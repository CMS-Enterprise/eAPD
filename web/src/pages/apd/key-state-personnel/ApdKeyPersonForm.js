import { TextField, ChoiceList } from '@cmsgov/design-system';
import PropTypes from 'prop-types';
import React, { 
  forwardRef,
  Fragment,
  useEffect,
  useReducer
} from 'react';
import { useForm, Controller } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/joi';
import { connect } from 'react-redux';

import { titleCase } from 'title-case';
import { t } from '../../../i18n';

import DollarField from '../../../components/DollarField';
import Dollars from '../../../components/Dollars';
import NumberField from '../../../components/NumberField';

import Joi from 'joi';
import { saveKeyPersonnel } from '../../../actions/editApd';

const getCheckedValue = value => {
  if (value !== null) {
    if (value === true) return 'yes';
    if (value === false) return 'no';
    return value;
  }
  return null;
};

const keyPersonSchema = Joi.object({
  name: Joi.string().required().messages({
    'string.base': 'Name is required',
    'string.empty': 'Name is required'
  }),
  email: Joi.string().required().messages({
    'string.base': 'Email is required',
    'string.empty': 'Email is required'
  }),
  position: Joi.string().required().messages({
    'string.base': 'Role is required',
    'string.empty': 'Role is required'
  }),
  hasCosts: Joi.string().required().messages({
    'string.base': 'Must select hourly or yearly.',
    'string.empty': 'Must select hourly or yearly.'
  }),
  costs: Joi.alternatives().conditional('hasCosts', {
    is: 'yes',
    then: Joi.object().pattern(
      /\d{4}/,
      Joi.number().positive().required().messages({
        'number.base': 'Provide a cost with benefits.',
        'number.empty': 'Provide a cost with benefits.',
        'number.format': 'Costs with Benefits should not be less than 0.',
        'number.positive': 'Costs with Benefits should not be less than 0.'
      })
    ),
    otherwise: Joi.any()
  }),
  fte: Joi.alternatives().conditional('hasCosts', {
    is: 'yes',
    then: Joi.object().pattern(
      /\d{4}/,
      Joi.number().positive().required().messages({
        'number.base': 'Provide an FTE.',
        'number.empty': 'Provide an FTE.',
        'number.format': 'FTE should not be less than 0.',
        'number.positive': 'FTE should not be less than 0.'
      })
    ),
    otherwise: Joi.any()
  })
})

const tRoot = 'apd.stateProfile.keyPersonnel';

const PersonForm = forwardRef(({ index, item, savePerson, years, setFormValid }, ref) => {
  PersonForm.displayName = 'PersonForm';
  const {
    handleSubmit,
    control,
    formState: { errors, isValid },
    resetField: resetFieldErrors
  } = useForm({
    defaultValues: {
      name: item.name,
      email: item.email,
      position: item.position,
      hasCosts: getCheckedValue(item.hasCosts),
      costs: item.costs,
      fte: item.fte
    },
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    resolver: joiResolver(keyPersonSchema)
  });

  useEffect(() => {
    console.log({ isValid });
    setFormValid(isValid);
  }, [isValid]); // eslint-disable-line react-hooks/exhaustive-deps

  const initialState = item;

  function reducer(state, action) {
    switch (action.type) {
      case 'updateField':
        return {
          ...state,
          [action.field]: action.payload
        };
      case 'updateCosts':
        return {
          ...state,
          costs: {
            ...state.costs,
            [action.year]: action.value
          }
        };
      case 'updateFte':
        return {
          ...state,
          fte: {
            ...state.fte,
            [action.year]: action.value
          }
        };
      default:
        throw new Error(
          'Unrecognized action type provided to ApdKeyPersonForm reducer'
        );
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState);

  const handleNameChange = e => {
    dispatch({
      type: 'updateField',
      field: 'name',
      payload: e.target.value
    });
  };

  const handleEmailChange = e => {
    dispatch({
      type: 'updateField',
      field: 'email',
      payload: e.target.value
    });
  };

  const handlePositionChange = e => {
    dispatch({
      type: 'updateField',
      field: 'position',
      payload: e.target.value
    });
  };

  const handleHasCostsChange = e => {
    let value = e.target.value === "yes" ? true : false;
    dispatch({
      type: 'updateField',
      field: 'hasCosts',
      payload: value
    });
  };

  const handleCostChange = (year, e) => {
    dispatch({
      type: 'updateCosts',
      year: year,
      value: e.target.value
    });
  };

  const handleFteChange = (year, e) => {
    dispatch({
      type: 'updateFte',
      year: year,
      value: e.target.value
    })
  }

  const onSubmit = e => {
    e.preventDefault();
    savePerson(index, state);
    handleSubmit(e);
  };

  const totalCostFte = year => {
    let total = state.costs[year] * state.fte[year];
    return total;
  }

  const primary = index === 0;

  return (
    <form index={index} onSubmit={onSubmit} aria-label="form">
      <h4 className="ds-h4">
        {primary
          ? titleCase(t(`${tRoot}.labels.titlePrimary`))
          : titleCase(t(`${tRoot}.labels.titleSecondary`))}
      </h4>
      <p className="ds-u-margin-bottom--0">
        {primary
          ? t(`${tRoot}.labels.notePrimary`)
          : t(`${tRoot}.labels.noteSecondary`)}
      </p>
      <Controller
        name="name"
        control={control}
        render={({ field: { onChange, ...props } }) => (
          <TextField
            {...props}
            id={`apd-state-profile-pocname${index}`}
            label={t(`${tRoot}.labels.name`)}
            data-cy={`key-person-${index}__name`}
            onChange={e => {
              handleNameChange(e);
              onChange(e);
            }}
            errorMessage={errors?.name?.message}
            errorPlacement="bottom"
          />
        )}
      />
      <Controller
        name="email"
        control={control}
        render={({ field: { onChange, ...props } }) => (
          <TextField
            {...props}
            id={`apd-state-profile-pocemail${index}`}
            label={t(`${tRoot}.labels.email`)}
            data-cy={`key-person-${index}__email`}
            onChange={e => {
              handleEmailChange(e);
              onChange(e);
            }}
            errorMessage={errors?.email?.message}
            errorPlacement="bottom"
          />
        )}
      />
      <Controller
        name="position"
        control={control}
        render={({ field: { onChange, ...props } }) => (
          <TextField
            {...props}
            id={`apd-state-profile-pocposition${index}`}
            label={t(`${tRoot}.labels.position`)}
            data-cy={`key-person-${index}__position`}
            onChange={e => {
              handlePositionChange(e);
              onChange(e);
            }}
            errorMessage={errors?.position?.message}
            errorPlacement="bottom"
          />
        )}
      />
      <Controller
        name="hasCosts"
        control={control}
        render={({
          field: { name, onChange, onBlur: hasCostsOnBlur, value }
        }) => (
          <ChoiceList
            label={t(`${tRoot}.labels.hasCosts`)}
            id={`apd-state-profile-hascosts${index}`}
            name={name}
            choices={[
              {
                label: "Yes",
                value: "yes",
                checked: value === "yes",
                checkedChildren: (
                  <div>
                    {years.map((year, i) => (
                      <Fragment key={year}>
                        <h5 className="ds-h5">FFY {year} Cost</h5>
                        <div className="ds-c-choice__checkedChild ds-u-padding-y--0">
                          <Controller
                            key={`${year}-cost`}
                            name={`costs.${year}`}
                            control={control}
                            render={({ field: { onChange: costOnChange, ...props } }) => (
                              <DollarField
                                {...props}
                                label="Cost with benefits"
                                size="medium"
                                data-cy={`key-person-${index}-${i}__cost`}
                                onChange={e => {
                                  handleCostChange(year, e);
                                  costOnChange(e);
                                }}
                              />
                            )}
                          />
                          <div>
                            <Fragment>
                              {errors?.costs && errors.costs[year] && (
                                <span className="ds-c-inline-error ds-c-field__error-message"
                                role="alert">
                                  {errors.costs[year]?.message}
                                </span>
                              )}
                            </Fragment>
                          </div>
                          <Controller
                            key={`${year}-fte`}
                            name={`fte.${year}`}
                            control={control}
                            render={({ field: { onChange: fteOnChange, ...props } }) => (
                              <NumberField
                                {...props}
                                label="FTE Allocation"
                                size="medium"
                                min={0}
                                hint="For example: 0.5 = 0.5 FTE = 50% time"
                                data-cy={`key-person-${index}-${i}__fte`}
                                onChange={e => {
                                  handleFteChange(year, e);
                                  fteOnChange(e);
                                }}
                              />
                            )}
                          />
                          <div>
                            <Fragment>
                              {errors?.fte && errors.fte[year] && (
                                <span className="ds-c-inline-error ds-c-field__error-message"
                                role="alert">
                                  {errors.fte[year]?.message}
                                </span>
                              )}
                            </Fragment>
                          </div>
                          <strong>Total: </strong>
                            <Dollars>{totalCostFte(year)}</Dollars>
                        </div>
                      </Fragment>
                    ))}
                  </div>
                )
              },
              {
                label: 'No',
                value: 'no',
                checked: value === 'no'
              }
            ]}
            type="radio"
            onChange={e => {
              resetFieldErrors('costs');
              resetFieldErrors('fte');
              handleHasCostsChange(e);
              onChange(e);
            }}
            onBlur={hasCostsOnBlur}
            onComponentBlur={hasCostsOnBlur}
            errorMessage={errors?.hasCosts?.message}
            errorPlacement="bottom"
          />
        )}
      />
      <input
        className="ds-u-visibility--hidden"
        type="submit"
        ref={ref}
        hidden
      />
    </form>
  );
});
PersonForm.propTypes = {
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    costs: PropTypes.object.isRequired,
    email: PropTypes.string.isRequired,
    hasCosts: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    fte: PropTypes.object.isRequired,
    position: PropTypes.string.isRequired
  }).isRequired,
  years: PropTypes.array.isRequired,
  savePerson: PropTypes.func.isRequired,
  setFormValid: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  savePerson: saveKeyPersonnel
};

export default connect(null, mapDispatchToProps, null, { forwardRef: true })(
  PersonForm
);

export { PersonForm as plain, mapDispatchToProps };
