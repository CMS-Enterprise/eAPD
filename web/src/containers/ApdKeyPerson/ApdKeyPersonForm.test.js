import { shallow } from 'enzyme';
import React from 'react';

import ApdKeyPersonForm from './ApdKeyPersonForm';

describe('the ApdKeyPersonForm component', () => {
  const props = {
    handleChange: jest.fn(),
    handleYearChange: jest.fn(),
    index: 1,
    item: {
      costs: {
        '1992': 100,
        '1993': 300
      },
      email: 'email address',
      hasCosts: true,
      key: 'person key',
      name: 'Bob the Builder',
      percentTime: '32',
      position: 'The Builder'
    },
    years: ['1992', '1993']
  };

  const component = shallow(<ApdKeyPersonForm {...props} />);

  beforeEach(() => {
    props.handleChange.mockClear();
    props.handleYearChange.mockClear();
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders correctly if the person does not have costs', () => {
    expect(
      shallow(
        <ApdKeyPersonForm
          {...props}
          item={{ ...props.item, hasCosts: false }}
        />
      )
    ).toMatchSnapshot();
  });

  describe('events', () => {
    [
      ['apd-state-profile-pocname1', 'name'],
      ['apd-state-profile-pocemail1', 'email'],
      ['apd-state-profile-pocposition1', 'position'],
      ['apd-state-profile-pocpercentTime1', 'percentTime']
    ].forEach(([formName, propName]) => {
      it(`handles changing the ${propName}`, () => {
        component
          .findWhere(c => c.prop('name') === formName)
          .simulate('change', { target: { value: 'new value' } });

        expect(props.handleChange).toHaveBeenCalledWith(
          1,
          propName,
          'new value',
          false
        );
      });
    });

    it('handles toggling hasCosts off', () => {
      component
        .findWhere(c => c.name() === 'ChoiceComponent' && c.prop('value') === 'no')
        .simulate('change');
      expect(props.handleChange).toHaveBeenCalledWith(
        1,
        'hasCosts',
        false,
        true
      );
    });

    it('handles toggling hasCosts on', () => {
      component
        .findWhere(c => c.name() === 'ChoiceComponent' && c.prop('value') === 'yes')
        .simulate('change');
      expect(props.handleChange).toHaveBeenCalledWith(
        1,
        'hasCosts',
        true,
        true
      );
    });

    it('handles changing cost for FFY', () => {
      const hasCostsForm = shallow(
        component
          .findWhere(c => c.name() === 'ChoiceComponent' && c.prop('value') === 'yes')
          .prop('checkedChildren')
      );

      hasCostsForm
        .find('DollarField')
        .first()
        .simulate('change', { target: { value: 9000 } });

      expect(props.handleYearChange).toHaveBeenCalledWith(1, '1992', 9000);
    });
  });
});
