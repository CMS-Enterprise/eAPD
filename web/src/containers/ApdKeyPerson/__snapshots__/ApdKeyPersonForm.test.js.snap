// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ApdKeyPersonForm component renders correctly 1`] = `
<Fragment>
  <h4
    className="ds-h4"
  >
    Additional
     APD Point of Contact
  </h4>
  <TextField
    autoFocus="true"
    label="Name"
    name="apd-state-profile-pocname1"
    onChange={[Function]}
    type="text"
    value="Bob the Builder"
  />
  <TextField
    label="Email"
    name="apd-state-profile-pocemail1"
    onChange={[Function]}
    type="text"
    value="email address"
  />
  <TextField
    label="Role"
    name="apd-state-profile-pocposition1"
    onChange={[Function]}
    type="text"
    value="The Builder"
  />
  <PercentField
    label="Percent commitment to program"
    name="apd-state-profile-pocpercentTime1"
    onChange={[Function]}
    size="small"
    value="32"
  />
  <fieldset
    className="ds-c-fieldset"
  >
    <legend
      className="ds-c-label"
    >
      Does this person have costs directly attributable to this program?
    </legend>
    <ChoiceComponent
      checked={false}
      name="apd-state-profile-hascosts-no1"
      onChange={[Function]}
      type="radio"
      value="no"
    >
      No
    </ChoiceComponent>
    <ChoiceComponent
      checked={true}
      checkedChildren={
        <div
          className="ds-c-choice__checkedChild ds-l-form-row"
        >
          <div
            className="ds-l-col--auto"
          >
            <DollarField
              label="FFY 1992"
              name="apd-state-profile-costs1-fy1992"
              onBlur={null}
              onChange={[Function]}
              size="small"
              value={100}
            />
          </div>
          <div
            className="ds-l-col--auto"
          >
            <DollarField
              label="FFY 1993"
              name="apd-state-profile-costs1-fy1993"
              onBlur={null}
              onChange={[Function]}
              size="small"
              value={300}
            />
          </div>
          <div
            className="ds-l-col--auto"
          >
            <FormLabel
              component="label"
            >
              Total
            </FormLabel>
            <div
              className="form--computed-value__input-aligned"
            >
              <Dollars
                long={false}
              >
                400
              </Dollars>
            </div>
          </div>
        </div>
      }
      name="apd-state-profile-hascosts-yes1"
      onChange={[Function]}
      type="radio"
      value="yes"
    >
      Yes
    </ChoiceComponent>
  </fieldset>
</Fragment>
`;

exports[`the ApdKeyPersonForm component renders correctly if the person does not have costs 1`] = `
<Fragment>
  <h4
    className="ds-h4"
  >
    Additional
     APD Point of Contact
  </h4>
  <TextField
    autoFocus="true"
    label="Name"
    name="apd-state-profile-pocname1"
    onChange={[Function]}
    type="text"
    value="Bob the Builder"
  />
  <TextField
    label="Email"
    name="apd-state-profile-pocemail1"
    onChange={[Function]}
    type="text"
    value="email address"
  />
  <TextField
    label="Role"
    name="apd-state-profile-pocposition1"
    onChange={[Function]}
    type="text"
    value="The Builder"
  />
  <PercentField
    label="Percent commitment to program"
    name="apd-state-profile-pocpercentTime1"
    onChange={[Function]}
    size="small"
    value="32"
  />
  <fieldset
    className="ds-c-fieldset"
  >
    <legend
      className="ds-c-label"
    >
      Does this person have costs directly attributable to this program?
    </legend>
    <ChoiceComponent
      checked={true}
      name="apd-state-profile-hascosts-no1"
      onChange={[Function]}
      type="radio"
      value="no"
    >
      No
    </ChoiceComponent>
    <ChoiceComponent
      checked={false}
      checkedChildren={
        <div
          className="ds-c-choice__checkedChild ds-l-form-row"
        >
          <div
            className="ds-l-col--auto"
          >
            <DollarField
              label="FFY 1992"
              name="apd-state-profile-costs1-fy1992"
              onBlur={null}
              onChange={[Function]}
              size="small"
              value={100}
            />
          </div>
          <div
            className="ds-l-col--auto"
          >
            <DollarField
              label="FFY 1993"
              name="apd-state-profile-costs1-fy1993"
              onBlur={null}
              onChange={[Function]}
              size="small"
              value={300}
            />
          </div>
          <div
            className="ds-l-col--auto"
          >
            <FormLabel
              component="label"
            >
              Total
            </FormLabel>
            <div
              className="form--computed-value__input-aligned"
            >
              <Dollars
                long={false}
              >
                400
              </Dollars>
            </div>
          </div>
        </div>
      }
      name="apd-state-profile-hascosts-yes1"
      onChange={[Function]}
      type="radio"
      value="yes"
    >
      Yes
    </ChoiceComponent>
  </fieldset>
</Fragment>
`;
